include/material/norsand.tcc

```diff
@@ -0,0 +1,693 @@
+
+  // Compute J3
+  double j3 = (dev_stress(0) * dev_stress(1) * dev_stress(2)) -
+              (dev_stress(2) * std::pow(dev_stress(3), 2));
+  if (Tdim == 3)
+    j3 += ((2 * dev_stress(3) * dev_stress(4) * dev_stress(5)) -
+           (dev_stress(0) * std::pow(dev_stress(4), 2)) -
+           (dev_stress(1) * std::pow(dev_stress(5), 2)));
+  (*state_vars).at("j3") = j3;
+
+  // Compute q
+  double deviatoric_q = std::sqrt(3 * j2);
+  if (deviatoric_q < 1.0E-15) deviatoric_q = 1.0E-15;
+  (*state_vars).at("q") = deviatoric_q;
+
+  // Compute Lode angle value
+  double lode_angle_val = (3. * std::sqrt(3.) / 2.) * (j3 / std::pow(j2, 1.5));
+  if (lode_angle_val > 1.0) lode_angle_val = 1.0;
+  if (lode_angle_val < -1.0) lode_angle_val = -1.0;
+
+  // Compute Lode angle (sin convention)
+  double lode_angle = (1. / 3.) * asin(lode_angle_val);
+  if (lode_angle > M_PI / 6.) lode_angle = M_PI / 6.;
+  if (lode_angle < -M_PI / 6.) lode_angle = -M_PI / 6.;
+  (*state_vars)["lode_angle"] = lode_angle;
+
+  // Compute M_theta (Jefferies and Shuttle, 2011)
+  const double cos_lode_angle = cos(3. / 2. * lode_angle + M_PI / 4.);
+  double M_theta = Mtc_ - std::pow(Mtc_, 2) / (3. + Mtc_) * cos_lode_angle;
+  (*state_vars)["M_theta"] = M_theta;
+
+  return true;
+}
+
+//! Compute state parameters
+template <unsigned Tdim>
+bool mpm::NorSand<Tdim>::compute_state_variables(const Vector6d& stress,
+                                                 const Vector6d& dstrain,
+                                                 mpm::dense_map* state_vars,
+                                                 FailureState yield_type) {
+
+  // Get state variables
+  const double mean_p = (*state_vars).at("p");
+  const double deviatoric_q = (*state_vars).at("q");
+  const double M_theta = (*state_vars).at("M_theta");
+  const double p_cohesion = (*state_vars).at("p_cohesion");
+  const double p_dilation = (*state_vars).at("p_dilation");
+  double p_image;
+  double e_image;
+
+  if (yield_type == FailureState::Elastic) {
+    // Keep the same pressure image and void ratio image at critical state
+    p_image = (*state_vars).at("p_image");
+    e_image = (*state_vars).at("e_image");
+  } else {
+    // Compute and update pressure image
+    if (bond_model_) {
+      p_image =
+          (mean_p + p_cohesion) *
+              std::pow(
+                  ((1 - N_ / M_theta * deviatoric_q / (mean_p + p_cohesion)) /
+                   (1 - N_)),
+                  ((N_ - 1) / N_)) -
+          p_cohesion - p_dilation;
+    } else {
+      p_image =
+          mean_p *
+          std::pow(((1 - N_ / M_theta * deviatoric_q / mean_p) / (1 - N_)),
+                   ((N_ - 1) / N_));
+    }
+    (*state_vars).at("p_image") = p_image;
+
+    // Compute and update void ratio image
+    // e_image = e_max_ - (e_max_ - e_min_) / log(crushing_pressure_ / p_image);
+    e_image = gamma_ - lambda_ * log(p_image / reference_pressure_);
+
+    if (e_image < 1.0E-15) e_image = 1.0E-15;
```

Comment:
In the current formulation, I think it's OK to have zero even though it's not physical. Previously, since I used logarithmic function, I tried to prevent zero.